lib/ansible/plugins/vars/host_group_vars.py

```diff
@@ -55,60 +55,92 @@
 
 import os
 from ansible.errors import AnsibleParserError
-from ansible.module_utils.common.text.converters import to_bytes, to_native, to_text
+from ansible.module_utils.common.text.converters import to_bytes, to_native
 from ansible.plugins.vars import BaseVarsPlugin
-from ansible.inventory.host import Host
-from ansible.inventory.group import Group
+from ansible.utils.path import basedir
 from ansible.utils.vars import combine_vars
 
+CANONICAL_PATHS = {}
 FOUND = {}  # type: dict[str, list[str]]
+NAK = set()
+HOSTS = set()
+GROUPS = set()
+
+
+def load_found_files(loader, data, found_files):
+    for found in found_files:
+        new_data = loader.load_from_file(found, cache=True, unsafe=True)
+        if new_data:  # ignore empty files
+            data = combine_vars(data, new_data)
+    return data
 
 
 class VarsModule(BaseVarsPlugin):
 
     REQUIRES_ENABLED = True
+    reuse_instance = True
 
     def get_vars(self, loader, path, entities, cache=True):
         ''' parses the inventory file '''
 
         if not isinstance(entities, list):
             entities = [entities]
 
-        super(VarsModule, self).get_vars(loader, path, entities)
+        # realpath is expensive
+        if path not in CANONICAL_PATHS:
+            CANONICAL_PATHS[path] = os.path.realpath(basedir(path))
+        realpath_basedir = CANONICAL_PATHS[path]
```

Comment:
This is a great place to use walrus, since we know this is a really hot code path- you can avoid consulting the path cache twice:

```suggestion
        # realpath is expensive
        if not (realpath_basedir := CANONICAL_PATHS.get(path)):
            CANONICAL_PATHS[path] = realpath_basedir = os.path.realpath(basedir(path))
```

Reply:
Adding the `get` here seems to be more expensive (using the test case from #79687), so I removed the extra path cache consultation by accessing the key immediately and handling the KeyError instead. Once the key is cached it makes sense that blindly accessing would be the most efficient.

<details>
<summary>some testing</summary>
<br>

```python
>>> def approach_1(container):
...     if 'something' not in container:
...         container['something'] = 'value'
...     value = container['something']
... 
>>> def approach_2(container):
...     if not (value := container.get('something')):
...         container['something'] = value = 'value'
... 
>>> def approach_3(container):
...     try:
...         value = container['something']
...     except KeyError:
...         container['something'] = value = 'value'
... 
>>>
>>> container = {i: "value" for i in range(5000)}
>>> timeit.timeit('approach_1(container)', globals=globals())
0.04540829104371369
>>>
>>> container = {i: "value" for i in range(5000)}
>>> timeit.timeit('approach_2(container)', globals=globals())
0.055584517074748874
>>>
>>> container = {i: "value" for i in range(5000)}
>>> timeit.timeit('approach_3(container)', globals=globals())
0.04524283786304295
>>>
>>>
>>> def exacerbate(func, number):
...     def wrapper(container):
...         for i in range(number):
...             func(container)
...     return wrapper
... 
>>>
>>> container = {i: "value" for i in range(10)}
>>> cProfile.run('exacerbate(approach_1, 5000)(container)')
         5005 function calls in 0.003 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5000    0.001    0.000    0.001    0.000 <stdin>:1(approach_1)
        1    0.000    0.000    0.000    0.000 <stdin>:1(exacerbate)
        1    0.002    0.002    0.003    0.003 <stdin>:2(wrapper)
        1    0.000    0.000    0.003    0.003 <string>:1(<module>)
        1    0.000    0.000    0.003    0.003 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


>>> container = {i: "value" for i in range(10)}
>>> cProfile.run('exacerbate(approach_2, 5000)(container)')
         10005 function calls in 0.004 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5000    0.002    0.000    0.002    0.000 <stdin>:1(approach_2)
        1    0.000    0.000    0.000    0.000 <stdin>:1(exacerbate)
        1    0.002    0.002    0.004    0.004 <stdin>:2(wrapper)
        1    0.000    0.000    0.004    0.004 <string>:1(<module>)
        1    0.000    0.000    0.004    0.004 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}
     5000    0.001    0.000    0.001    0.000 {method 'get' of 'dict' objects}


>>> container = {i: "value" for i in range(10)}
>>> cProfile.run('exacerbate(approach_3, 5000)(container)')
         5005 function calls in 0.001 seconds

   Ordered by: standard name

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     5000    0.000    0.000    0.000    0.000 <stdin>:1(approach_3)
        1    0.000    0.000    0.000    0.000 <stdin>:1(exacerbate)
        1    0.000    0.000    0.001    0.001 <stdin>:2(wrapper)
        1    0.000    0.000    0.001    0.001 <string>:1(<module>)
        1    0.000    0.000    0.001    0.001 {built-in method builtins.exec}
        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}


>>>
>>>
>>> container = {i: "value" for i in range(10)}
>>> timeit.timeit('exacerbate(approach_1, 5000)(container)', globals=globals())
255.56839252007194
>>>
>>> container = {i: "value" for i in range(10)}
>>> timeit.timeit('exacerbate(approach_2, 5000)(container)', globals=globals())
263.8696411601268
>>>
>>> container = {i: "value" for i in range(10)}
>>> timeit.timeit('exacerbate(approach_3, 5000)(container)', globals=globals())
213.75293880398385
```
</details>